runtime/cudaq/algorithms/evolve_internal.h

```diff
@@ -0,0 +1,168 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include "common/EvolveResult.h"
+#include "common/KernelWrapper.h"
+#include "cudaq/algorithms/get_state.h"
+#include "cudaq/host_config.h"
+#include "cudaq/operators.h"
+#include "cudaq/platform.h"
+#include "cudaq/platform/QuantumExecutionQueue.h"
+#include "cudaq/schedule.h"
+
+namespace cudaq {
+class base_integrator;
+
+/// @brief Return type for asynchronous `evolve_async`.
+using async_evolve_result = std::future<evolve_result>;
+
+namespace __internal__ {
+// Internal methods for evolve implementation on circuit simulators.
+
+/// @brief Evolve from an initial state to the final state, no intermediate
+/// states.
+template <typename QuantumKernel>
+evolve_result evolve(state initial_state, QuantumKernel &&kernel,
+                     const std::vector<spin_op> &observables = {},
+                     int shots_count = -1) {
+  state final_state =
+      get_state(std::forward<QuantumKernel>(kernel), initial_state);
+  if (observables.size() == 0)
+    return evolve_result(final_state);
+
+  auto prepare_state = [final_state]() { auto qs = qvector<2>(final_state); };
+  std::vector<observe_result> final_expectations;
+  for (auto observable : observables) {
+    shots_count <= 0
+        ? final_expectations.push_back(observe(prepare_state, observable))
+        : final_expectations.push_back(
+              observe(shots_count, prepare_state, observable));
+  }
+  return evolve_result(final_state, final_expectations);
+}
+
+/// @brief Evolve from an initial state to the final state and gather
+/// intermediate states.
+// Step evolution is provided as `kernels`.
+template <typename QuantumKernel>
+evolve_result evolve(state initial_state, std::vector<QuantumKernel> kernels,
```

Comment:
is it intended that you only allow a single type of QuantumKernel for each call here to evolve? If so, should this be `const std::vector<QuantumKernel>& kernels`?